kotest-framework/kotest-framework-multiplatform-plugin-gradle/src/main/kotlin/io/kotest/framework/multiplatform/gradle/KotestMultiplatformCompilerGradlePlugin.kt

```diff
@@ -1,105 +1,93 @@
-    * specified, then defaults to using the same version as the engine dependency.
+    * In [isApplicable] a [Project] instance is available, so fetch the extension 'normally'. This helps ensure the
+    * Gradle API is used correctly.
     */
-   private val version: String? by lazy {
-      val versionFromExtension = extension?.compilerPluginVersion?.orNull
-      if (versionFromExtension != null) {
-         println("Kotest compiler plugin [$versionFromExtension]")
-         return@lazy versionFromExtension
-      }
-
-      val engineDep = engineDeps().firstOrNull() ?: error("Cannot determine Kotest compiler plugin version if no Kotest engine dependencies are present")
-      val version = engineDep.version ?: return@lazy null
-
-      if (version.contains("LOCAL")) {
-         println("Detected dev engine version [$version]")
-      }
+   private var kotestExtension: KotestPluginExtension? = null
 
-      return@lazy version
+   override fun apply(target: Project) {
+      kotestExtension = target.createKotestExtension()
    }
 
-   private fun engineDeps(): List<Dependency> {
-      val project = target ?: error(missingProjectValError)
-
-      return project.configurations
-         .flatMap { it.all }
-         .flatMap { it.dependencies }
-         .filter { it.group == KotestGroupId && it.name.startsWith(engineDepPrefix) }
+   private fun Project.createKotestExtension(): KotestPluginExtension {
+      return extensions.create<KotestPluginExtension>(kotestPluginExtensionName).apply {
+         kotestCompilerPluginVersion.convention(KOTEST_COMPILER_PLUGIN_VERSION)
+      }
    }
 
    override fun getCompilerPluginId() = compilerPluginId
 
    override fun getPluginArtifact(): SubpluginArtifact =
-      SubpluginArtifact(KotestGroupId, KotestEmbeddableCompilerArtifactId, version)
+      SubpluginArtifact(
+         KotestGroupId,
+         KotestEmbeddableCompilerArtifactId,
+         kotestExtension?.kotestCompilerPluginVersion?.orNull
+      )
 
    // This will soon be deprecated and removed, see https://youtrack.jetbrains.com/issue/KT-51301.
    override fun getPluginArtifactForNative(): SubpluginArtifact =
-      SubpluginArtifact(KotestGroupId, KotestNativeArtifactId, version)
+      SubpluginArtifact(KotestGroupId, KotestNativeArtifactId, kotestExtension?.kotestCompilerPluginVersion?.orNull)
 
    override fun isApplicable(kotlinCompilation: KotlinCompilation<*>): Boolean {
+      val project = kotlinCompilation.target.project
+      val kotestExtension = project.extensions.findByType<KotestPluginExtension>()
+
       return when {
-         // if we can't find a version to use then we won't apply to this module
-         engineDeps().isEmpty() -> {
-            println("Warning: Kotest plugin has been added to project $target, but the project does not contain a Kotest engine dependency. Kotest will not be enabled.")
+         kotestExtension == null                                -> {
+            // this shouldn't happen
```

Comment:
If this shouldn't happen, should we just throw an exception and give up?

Reply:
If there's a bug (perhaps the Kotlin compiler plugin is not created in the correct order, or the Kotest plugin isn't applied correctly), then personally I'd rather an warning message so I can continue without Kotest tests than an exception that completely blocks all development.

Reply:
I'm not sure I agree - if there's something wrong setting up Kotest, I'd want it to fail early and loud, otherwise there's a risk Gradle is silently not running the tests and the build is passing when in fact there are failing tests.

Reply:
Yeah I think we'd want to fail, the user can remove the kotest plugin if there's no workaround for them.

Reply:
Good point, done!